src/matcher_combinators/midje.clj

```diff
@@ -25,6 +25,29 @@
       (checking/as-data-laden-falsehood
         {:notes [(str "Input wasn't a matcher: " matcher)]}))))
 
+(defn var->qualified-ref [datatype]
+  (symbol (str
+            (-> parser/type-map datatype meta :ns ns-name)
+            "/"
+            (-> parser/type-map datatype meta :name))))
+
+(defmacro match-with [type->matcher expected-matcher]
+  (let [dispatch-vars+matcher-targets (mapcat
+                                        (fn [[k v]] [(var->qualified-ref k) v])
+                                        type->matcher)]
+    `(checkers.defining/checker [actual#]
+       (with-redefs [~@dispatch-vars+matcher-targets]
+          (if (core/matcher? ~expected-matcher)
+            (check-match ~expected-matcher actual#)
+            (checking/as-data-laden-falsehood
+              {:notes [(str "Input wasn't a matcher: " ~expected-matcher)]}))))))
+
+(checkers.defining/defchecker match-equals [matcher]
+  (match-with {:map (fn [exp] (core/->EqualsMap exp))} matcher))
+
+(checkers.defining/defchecker match-roughly [delta matcher]
+  (match-with {:number (fn [exp] (core/->Roughly delta exp))} matcher))
```

Comment:
Regardless of how it is implemented, `match-with` should give us a crazy amount of flexibility to meet various checker related feature requests found in `midje`'s issues page.

Reply:
Hmm, on thinking about this a little more, my approach will have extensibility issues because if someone wants to use `match-with` with a type not in the the `type-map` (https://github.com/nubank/matcher-combinators/pull/29/files#diff-51dee594687ff558df6b1fc75a4cc561R35) it won't currently work. Going to continue to mull this over; maybe something like multi-methods can save us here